platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaObjectSerializationCodec.kt

```diff
@@ -159,14 +165,25 @@ class JavaObjectSerializationCodec(
         override suspend fun WriteContext.encode(value: Any) {
             encodePreservingIdentityOf(value) {
                 val replacement = writeReplace.invoke(value)
-                if (replacement is SerializedLambda) {
-                    writeEnum(Format.SerializedLambda)
-                    SerializedLambdaParametersCheckingCodec.run {
-                        encode(replacement)
+                when {
+                    replacement is SerializedLambda -> {
+                        writeEnum(Format.SerializedLambda)
+                        SerializedLambdaParametersCheckingCodec.run {
+                            encode(replacement)
+                        }
+                    }
+
+                    replacement::class.java === value::class.java -> {
+                        // Avoid a StackOverflowException when the replacement and value are of the same type.
+                        // TODO:configuration-cache This is likely incorrect when the class also supports the `writeObject` protocol.
```

Comment:
❓ 

Can't we test for this? (as a `@ToBeImplemented` test, if confirmed)

Reply:
> Can't we test for this? (as a `@ToBeImplemented` test, if confirmed)

We can. I was saving that for another day.

Reply:
Maybe then expand this TODO to include the reason why you think this would be incorrect?

Reply:
Done.